p2p/proto/consensus.proto

```diff
@@ -1,42 +1,75 @@
 syntax = "proto3";
 
 import "p2p/proto/common.proto";
+import "p2p/proto/header.proto";
 import "p2p/proto/transaction.proto";
-import "p2p/proto/state.proto";
-import "google/protobuf/timestamp.proto";
-
 
 // WIP - will change
 
-message Proposal {
-    uint64                    block_number      = 2;
-    uint32                    round             = 3;
-    uint32                    pol               = 4;  // proof of lock
-    Hash                      block_header_hash = 5;
-    google.protobuf.Timestamp timestamp         = 6;
-    ConsensusSignature        signature         = 7;
-}
-
-// A block proposal is a series of (Transactions+, StateDiff)* BlockHeader
-
 message Vote {
-    enum  Type {
-        UNKNOWN   = 0;
-        Proposal  = 1;
-        Prevote   = 2;
-        Precommit = 3;
+    enum  VoteType {
+        Prevote   = 0;
+        Precommit = 1;
     };
 
-    Proposal           proposal          = 1;
-    Address            validator_address = 2;
-    int32              validator_index   = 3;  // ???
-    ConsensusSignature signature         = 4;
+    // We use a type field to distinguish between prevotes and precommits instead of different
+    // messages, to make sure the data, and therefore the signatures, are unambiguous between
+    // Prevote and Precommit.
+    VoteType           vote_type    = 1;
+    Hash               parent_hash  = 2;
+    uint64             block_number = 3;
+    uint64             fork_id      = 4;
+    uint32             round        = 5;
+    // This is optional since a vote can be NIL.
+    optional Hash      block_hash   = 6;
+    // Identifies the voter.
+    Address            voter        = 7;
+}
+
+message ProposalInit {
+    Hash    parent_hash    = 1;
+    uint64  block_number   = 2;
+    uint64  fork_id        = 3;
+    uint32  proposal_round = 4;
+    Address proposer       = 5;
 }
 
-message CreateBlock {
+// Produce the full Tendermint Proposal message once the proposer node has finished building it.
+// Tendermint just ID's the proposal's content, block_hash.
+message ProposalFin {
+    Hash            parent_hash    = 1;
```

Comment:
If ProposalFin needs to repeat fields form ProposalInit, why don't Transactions and BlockProof need to have them?

Reply:
The reason ProposalFin repeats the field is because we need the proposer to sign on all of these fields. I always want the signature on a message to be only on the content in the message. If I don't repeat fields then the consensus signature is on some fields from ProposalFin and some from ProposalInit. Updated the comment.

I think the structure here is more straightforward than adding a new "proposal id". The Proposal struct from Tendermint is well defined and we aren't expecting it to change.

Reply:
For what I want to achieve, no I do not need to repeat the fields in the context (transactions & proof). I want to have the ProposalFin include the full consensus signature on the proposal. I want the signature to match the content of the message. Therefore, only in ProposalFin do I need to repeat.

Reply:
Don't repeat. The proposer needs to sign on these, but no need to repeat. Nodes will need to keep the init message and confirm the fields are the same, so the proposer is not cheating (telling them to validate against one state and then finalizing as if it's another). 

And, the signature will need to include this information for transactions and proof as well to support message interleaving (in the rare case that a proposer proposes on several rounds and there are network issues so messages come out of band, especially since we user relays.